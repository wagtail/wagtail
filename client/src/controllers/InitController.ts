import { Controller } from '@hotwired/stimulus';
import { debounce } from '../utils/debounce';

/**
 * Adds the ability for a controlled element to add or remove classes
 * when ready to be interacted with.
 *
 * @example
 * <div class="keep-me hide-me" data-controller="w-init" data-w-init-remove-class="hide-me" data-w-init-ready-class="loaded">
 *   When the DOM is ready, this div will have the class 'loaded' added and 'hide-me' removed.
 * </div>
 */
export class InitController extends Controller<HTMLElement> {
  static classes = ['ready', 'remove'];

  static values = {
    delay: { default: -1, type: Number },
  };

  declare readonly readyClasses: string[];
  declare readonly removeClasses: string[];

  declare delayValue: number;

  connect() {
    this.ready();
  }

  /**
   * Add the ready classes and remove the remove classes after a delay.
   * By default, the action will be immediate (negative value).
   * Even when immediate, allow for a microtask delay to allow for other
   * controllers to connect, then do any updates do classes/dispatch events.
   */
  ready() {
    const delayValue = this.delayValue;

    debounce(() => true, delayValue < 0 ? null : delayValue)().then(() => {
      this.element.classList.add(...this.readyClasses);
      this.element.classList.remove(...this.removeClasses);
      this.dispatch('ready', { bubbles: true, cancelable: false });
      this.remove();
    });
  }

  /**
   * Allow the controller to remove itself as it's no longer needed when the init has completed.
   */
  remove() {
    const element = this.element;
    const controllerAttribute = this.application.schema.controllerAttribute;
    const controllers =
      element.getAttribute(controllerAttribute)?.split(' ') ?? [];
    const newControllers = controllers
      .filter((identifier) => identifier !== this.identifier)
      .join(' ');
    if (newControllers) {
      element.setAttribute(controllerAttribute, newControllers);
    } else {
      element.removeAttribute(controllerAttribute);
    }
  }
}
