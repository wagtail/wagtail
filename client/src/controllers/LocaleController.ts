import { Controller } from '@hotwired/stimulus';

/**
 * Localizes elements in the current locale.
 */
export class LocaleController extends Controller<HTMLSelectElement> {
  /**
   * Localize an IANA time zone in the current locale.
   *
   * @param timeZone An IANA time zone string
   * @param format Time zone name formatting option
   * @returns formatted time zone name in the current locale
   */
  static localizeTimeZone(
    timeZone: string,
    format: Intl.DateTimeFormatOptions['timeZoneName'],
  ) {
    const df = new Intl.DateTimeFormat(document.documentElement.lang, {
      timeZone,
      timeZoneName: format,
    });
    const parts = df.formatToParts(new Date());
    return parts.find((part) => part.type === 'timeZoneName')!.value;
  }

  /**
   *
   * @param timeZone An IANA time zone string
   * @returns formatted time zone name in the current locale with short and long
   * labels, e.g. `"GMT+7 (Western Indonesia Time)"`
   */
  static getTZLabel(timeZone: string) {
    const shortLabel = LocaleController.localizeTimeZone(timeZone, 'short');
    const longLabel = LocaleController.localizeTimeZone(timeZone, 'long');
    return `${shortLabel} (${longLabel})`;
  }

  /**
   * Localize the time zone `<options>` of a `<select>` element in the current
   * locale.
   */
  localizeTimeZoneOptions(
    event?: Event & { params?: { serverTimeZone?: string } },
  ) {
    const element = (event?.target as HTMLSelectElement) || this.element;
    const serverTimeZone = event?.params?.serverTimeZone;
    Array.from(element.options).forEach((opt) => {
      const timeZone = opt.value || serverTimeZone;
      if (!timeZone) return;
      const localized = LocaleController.getTZLabel(timeZone);
      const option = opt;
      option.textContent = `${option.textContent}: ${localized}`;
    });
  }
}
