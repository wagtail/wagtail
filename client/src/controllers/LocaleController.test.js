import { Application } from '@hotwired/stimulus';
import { LocaleController } from './LocaleController';
import { InitController } from './InitController';

describe('LocaleController', () => {
  let app;
  let select;

  const setup = async (html) => {
    document.body.innerHTML = `<main>${html}</main>`;
    select = document.querySelector('select');
    select.innerHTML = /* html */ `
      <option value="" selected>Use server time zone</option>
      <option value="Africa/Abidjan">Africa/Abidjan</option>
      <option value="America/Argentina/Jujuy">America/Argentina/Jujuy</option>
      <option value="America/Indiana/Knox">America/Indiana/Knox</option>
      <option value="Antarctica/Rothera">Antarctica/Rothera</option>
      <option value="Arctic/Longyearbyen">Arctic/Longyearbyen</option>
      <option value="Asia/Katmandu">Asia/Katmandu</option>
      <option value="Atlantic/Canary">Atlantic/Canary</option>
      <option value="Australia/South">Australia/South</option>
      <option value="Brazil/East">Brazil/East</option>
      <option value="Canada/Atlantic">Canada/Atlantic</option>
      <option value="Chile/Continental">Chile/Continental</option>
      <option value="EST">EST</option>
      <option value="Etc/GMT-7">Etc/GMT-7</option>
      <option value="Europe/Brussels">Europe/Brussels</option>
      <option value="GMT">GMT</option>
      <option value="Indian/Maldives">Indian/Maldives</option>
      <option value="Pacific/Tarawa">Pacific/Tarawa</option>
      <option value="UTC">UTC</option>
      <option value="Universal">Universal</option>
      <option value="Zulu">Zulu</option>
    `;

    app = Application.start();
    app.register('w-locale', LocaleController);
    app.register('w-init', InitController);

    await Promise.resolve();
  };

  afterEach(() => {
    app?.stop();
    jest.clearAllMocks();
  });

  describe('localizing time zone options', () => {
    it('should append localized time zone labels to the options', async () => {
      document.documentElement.lang = 'en-US';
      await setup(/* html */ `
        <select
          name="locale-current_time_zone"
          data-controller="w-init w-locale"
          data-action="w-init:ready->w-locale#localizeTimeZoneOptions"
          data-w-locale-server-time-zone-param="Europe/London"
        >
        </select>
      `);

      expect(select.getAttribute('data-controller')).toEqual('w-locale');
      const selected = select.selectedOptions[0];
      expect(selected).toBeTruthy();
      expect(selected.value).toEqual('');
      expect(selected.textContent).toEqual(
        'Use server time zone: GMT (Greenwich Mean Time)',
      );
      expect(select).toMatchSnapshot();
    });
  });

  it('should localize to the current HTML locale and use the server time zone param for the default', async () => {
    document.documentElement.lang = 'id-ID';
    await setup(/* html */ `
      <select
        name="locale-current_time_zone"
        data-controller="w-init w-locale"
        data-action="w-init:ready->w-locale#localizeTimeZoneOptions"
        data-w-locale-server-time-zone-param="Asia/Jakarta"
      >
      </select>
    `);

    expect(select.getAttribute('data-controller')).toEqual('w-locale');
    const selected = select.selectedOptions[0];
    expect(selected).toBeTruthy();
    expect(selected.value).toEqual('');
    expect(selected.textContent).toEqual(
      'Use server time zone: WIB (Waktu Indonesia Barat)',
    );
    expect(select).toMatchSnapshot();
  });

  it('should skip updating the default option if server time zone is not provided', async () => {
    document.documentElement.lang = 'ar';
    await setup(/* html */ `
      <select
        name="locale-current_time_zone"
        data-controller="w-init w-locale"
        data-action="w-init:ready->w-locale#localizeTimeZoneOptions"
      >
      </select>
    `);

    expect(select.getAttribute('data-controller')).toEqual('w-locale');
    const selected = select.selectedOptions[0];
    expect(selected).toBeTruthy();
    expect(selected.value).toEqual('');
    expect(selected.textContent).toEqual('Use server time zone');
    expect(select).toMatchSnapshot();
  });

  it('should allow updating the time zone options on an uncontrolled select element via events', async () => {
    document.documentElement.lang = 'id-ID';
    await setup(/* html */ `
      <form data-controller="w-locale">
        <select
          name="locale-current_time_zone"
          data-action="custom:event->w-locale#localizeTimeZoneOptions"
          data-w-locale-server-time-zone-param="Asia/Tokyo"
        >
        </select>
      </form>
    `);
    select.dispatchEvent(new CustomEvent('custom:event'));
    await Promise.resolve();

    expect(select.hasAttribute('data-controller')).toBe(false);
    const selected = select.selectedOptions[0];
    expect(selected).toBeTruthy();
    expect(selected.value).toEqual('');
    expect(selected.textContent).toEqual(
      'Use server time zone: GMT+9 (Waktu Standar Jepang)',
    );
    expect(select).toMatchSnapshot();
  });
});
