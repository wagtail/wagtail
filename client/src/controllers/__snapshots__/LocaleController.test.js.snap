// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleController localizing time zone options should append localized time zone labels to the options 1`] = `
<select
  data-action="w-init:ready->w-locale#localizeTimeZoneOptions"
  data-controller="w-locale"
  data-w-locale-server-time-zone-param="Europe/London"
  name="locale-current_time_zone"
>
  
      
  <option
    selected=""
    value=""
  >
    Use server time zone: GMT (Greenwich Mean Time)
  </option>
  
      
  <option
    value="Africa/Abidjan"
  >
    Africa/Abidjan: GMT (Greenwich Mean Time)
  </option>
  
      
  <option
    value="America/Argentina/Jujuy"
  >
    America/Argentina/Jujuy: GMT-3 (Argentina Standard Time)
  </option>
  
      
  <option
    value="America/Indiana/Knox"
  >
    America/Indiana/Knox: CST (Central Standard Time)
  </option>
  
      
  <option
    value="Antarctica/Rothera"
  >
    Antarctica/Rothera: GMT-3 (Rothera Time)
  </option>
  
      
  <option
    value="Arctic/Longyearbyen"
  >
    Arctic/Longyearbyen: GMT+1 (Central European Standard Time)
  </option>
  
      
  <option
    value="Asia/Katmandu"
  >
    Asia/Katmandu: GMT+5:45 (Nepal Time)
  </option>
  
      
  <option
    value="Atlantic/Canary"
  >
    Atlantic/Canary: GMT (Western European Standard Time)
  </option>
  
      
  <option
    value="Australia/South"
  >
    Australia/South: GMT+10:30 (Australian Central Daylight Time)
  </option>
  
      
  <option
    value="Brazil/East"
  >
    Brazil/East: GMT-3 (Brasilia Standard Time)
  </option>
  
      
  <option
    value="Canada/Atlantic"
  >
    Canada/Atlantic: AST (Atlantic Standard Time)
  </option>
  
      
  <option
    value="Chile/Continental"
  >
    Chile/Continental: GMT-3 (Chile Summer Time)
  </option>
  
      
  <option
    value="EST"
  >
    EST: EST (Eastern Standard Time)
  </option>
  
      
  <option
    value="Etc/GMT-7"
  >
    Etc/GMT-7: GMT+7 (GMT+07:00)
  </option>
  
      
  <option
    value="Europe/Brussels"
  >
    Europe/Brussels: GMT+1 (Central European Standard Time)
  </option>
  
      
  <option
    value="GMT"
  >
    GMT: UTC (Coordinated Universal Time)
  </option>
  
      
  <option
    value="Indian/Maldives"
  >
    Indian/Maldives: GMT+5 (Maldives Time)
  </option>
  
      
  <option
    value="Pacific/Tarawa"
  >
    Pacific/Tarawa: GMT+12 (Gilbert Islands Time)
  </option>
  
      
  <option
    value="UTC"
  >
    UTC: UTC (Coordinated Universal Time)
  </option>
  
      
  <option
    value="Universal"
  >
    Universal: UTC (Coordinated Universal Time)
  </option>
  
      
  <option
    value="Zulu"
  >
    Zulu: UTC (Coordinated Universal Time)
  </option>
  
    
</select>
`;

exports[`LocaleController should allow updating the time zone options on an uncontrolled select element via events 1`] = `
<select
  data-action="custom:event->w-locale#localizeTimeZoneOptions"
  data-w-locale-server-time-zone-param="Asia/Tokyo"
  name="locale-current_time_zone"
>
  
      
  <option
    selected=""
    value=""
  >
    Use server time zone: GMT+9 (Waktu Standar Jepang)
  </option>
  
      
  <option
    value="Africa/Abidjan"
  >
    Africa/Abidjan: GMT (Greenwich Mean Time)
  </option>
  
      
  <option
    value="America/Argentina/Jujuy"
  >
    America/Argentina/Jujuy: GMT-3 (Waktu Standar Argentina)
  </option>
  
      
  <option
    value="America/Indiana/Knox"
  >
    America/Indiana/Knox: CST (Waktu Standar Tengah)
  </option>
  
      
  <option
    value="Antarctica/Rothera"
  >
    Antarctica/Rothera: GMT-3 (Waktu Rothera)
  </option>
  
      
  <option
    value="Arctic/Longyearbyen"
  >
    Arctic/Longyearbyen: GMT+1 (Waktu Standar Eropa Tengah)
  </option>
  
      
  <option
    value="Asia/Katmandu"
  >
    Asia/Katmandu: GMT+5.45 (Waktu Nepal)
  </option>
  
      
  <option
    value="Atlantic/Canary"
  >
    Atlantic/Canary: GMT (Waktu Standar Eropa Barat)
  </option>
  
      
  <option
    value="Australia/South"
  >
    Australia/South: GMT+10.30 (Waktu Musim Panas Tengah Australia)
  </option>
  
      
  <option
    value="Brazil/East"
  >
    Brazil/East: GMT-3 (Waktu Standar Brasil)
  </option>
  
      
  <option
    value="Canada/Atlantic"
  >
    Canada/Atlantic: AST (Waktu Standar Atlantik)
  </option>
  
      
  <option
    value="Chile/Continental"
  >
    Chile/Continental: GMT-3 (Waktu Musim Panas Cile)
  </option>
  
      
  <option
    value="EST"
  >
    EST: EST (Waktu Standar Timur)
  </option>
  
      
  <option
    value="Etc/GMT-7"
  >
    Etc/GMT-7: GMT+7 (GMT+07.00)
  </option>
  
      
  <option
    value="Europe/Brussels"
  >
    Europe/Brussels: GMT+1 (Waktu Standar Eropa Tengah)
  </option>
  
      
  <option
    value="GMT"
  >
    GMT: UTC (Waktu Universal Terkoordinasi)
  </option>
  
      
  <option
    value="Indian/Maldives"
  >
    Indian/Maldives: GMT+5 (Waktu Maladewa)
  </option>
  
      
  <option
    value="Pacific/Tarawa"
  >
    Pacific/Tarawa: GMT+12 (Waktu Kep. Gilbert)
  </option>
  
      
  <option
    value="UTC"
  >
    UTC: UTC (Waktu Universal Terkoordinasi)
  </option>
  
      
  <option
    value="Universal"
  >
    Universal: UTC (Waktu Universal Terkoordinasi)
  </option>
  
      
  <option
    value="Zulu"
  >
    Zulu: UTC (Waktu Universal Terkoordinasi)
  </option>
  
    
</select>
`;

exports[`LocaleController should localize to the current HTML locale and use the server time zone param for the default 1`] = `
<select
  data-action="w-init:ready->w-locale#localizeTimeZoneOptions"
  data-controller="w-locale"
  data-w-locale-server-time-zone-param="Asia/Jakarta"
  name="locale-current_time_zone"
>
  
      
  <option
    selected=""
    value=""
  >
    Use server time zone: WIB (Waktu Indonesia Barat)
  </option>
  
      
  <option
    value="Africa/Abidjan"
  >
    Africa/Abidjan: GMT (Greenwich Mean Time)
  </option>
  
      
  <option
    value="America/Argentina/Jujuy"
  >
    America/Argentina/Jujuy: GMT-3 (Waktu Standar Argentina)
  </option>
  
      
  <option
    value="America/Indiana/Knox"
  >
    America/Indiana/Knox: CST (Waktu Standar Tengah)
  </option>
  
      
  <option
    value="Antarctica/Rothera"
  >
    Antarctica/Rothera: GMT-3 (Waktu Rothera)
  </option>
  
      
  <option
    value="Arctic/Longyearbyen"
  >
    Arctic/Longyearbyen: GMT+1 (Waktu Standar Eropa Tengah)
  </option>
  
      
  <option
    value="Asia/Katmandu"
  >
    Asia/Katmandu: GMT+5.45 (Waktu Nepal)
  </option>
  
      
  <option
    value="Atlantic/Canary"
  >
    Atlantic/Canary: GMT (Waktu Standar Eropa Barat)
  </option>
  
      
  <option
    value="Australia/South"
  >
    Australia/South: GMT+10.30 (Waktu Musim Panas Tengah Australia)
  </option>
  
      
  <option
    value="Brazil/East"
  >
    Brazil/East: GMT-3 (Waktu Standar Brasil)
  </option>
  
      
  <option
    value="Canada/Atlantic"
  >
    Canada/Atlantic: AST (Waktu Standar Atlantik)
  </option>
  
      
  <option
    value="Chile/Continental"
  >
    Chile/Continental: GMT-3 (Waktu Musim Panas Cile)
  </option>
  
      
  <option
    value="EST"
  >
    EST: EST (Waktu Standar Timur)
  </option>
  
      
  <option
    value="Etc/GMT-7"
  >
    Etc/GMT-7: GMT+7 (GMT+07.00)
  </option>
  
      
  <option
    value="Europe/Brussels"
  >
    Europe/Brussels: GMT+1 (Waktu Standar Eropa Tengah)
  </option>
  
      
  <option
    value="GMT"
  >
    GMT: UTC (Waktu Universal Terkoordinasi)
  </option>
  
      
  <option
    value="Indian/Maldives"
  >
    Indian/Maldives: GMT+5 (Waktu Maladewa)
  </option>
  
      
  <option
    value="Pacific/Tarawa"
  >
    Pacific/Tarawa: GMT+12 (Waktu Kep. Gilbert)
  </option>
  
      
  <option
    value="UTC"
  >
    UTC: UTC (Waktu Universal Terkoordinasi)
  </option>
  
      
  <option
    value="Universal"
  >
    Universal: UTC (Waktu Universal Terkoordinasi)
  </option>
  
      
  <option
    value="Zulu"
  >
    Zulu: UTC (Waktu Universal Terkoordinasi)
  </option>
  
    
</select>
`;

exports[`LocaleController should skip updating the default option if server time zone is not provided 1`] = `
<select
  data-action="w-init:ready->w-locale#localizeTimeZoneOptions"
  data-controller="w-locale"
  name="locale-current_time_zone"
>
  
      
  <option
    selected=""
    value=""
  >
    Use server time zone
  </option>
  
      
  <option
    value="Africa/Abidjan"
  >
    Africa/Abidjan: غرينتش (توقيت غرينتش)
  </option>
  
      
  <option
    value="America/Argentina/Jujuy"
  >
    America/Argentina/Jujuy: غرينتش-٣ (توقيت الأرجنتين الرسمي)
  </option>
  
      
  <option
    value="America/Indiana/Knox"
  >
    America/Indiana/Knox: غرينتش-٦ (التوقيت الرسمي المركزي لأمريكا الشمالية)
  </option>
  
      
  <option
    value="Antarctica/Rothera"
  >
    Antarctica/Rothera: غرينتش-٣ (توقيت روثيرا)
  </option>
  
      
  <option
    value="Arctic/Longyearbyen"
  >
    Arctic/Longyearbyen: غرينتش+١ (توقيت وسط أوروبا الرسمي)
  </option>
  
      
  <option
    value="Asia/Katmandu"
  >
    Asia/Katmandu: غرينتش+٥:٤٥ (توقيت نيبال)
  </option>
  
      
  <option
    value="Atlantic/Canary"
  >
    Atlantic/Canary: غرينتش (توقيت غرب أوروبا الرسمي)
  </option>
  
      
  <option
    value="Australia/South"
  >
    Australia/South: غرينتش+١٠:٣٠ (توقيت وسط أستراليا الصيفي)
  </option>
  
      
  <option
    value="Brazil/East"
  >
    Brazil/East: غرينتش-٣ (توقيت برازيليا الرسمي)
  </option>
  
      
  <option
    value="Canada/Atlantic"
  >
    Canada/Atlantic: غرينتش-٤ (التوقيت الرسمي الأطلسي)
  </option>
  
      
  <option
    value="Chile/Continental"
  >
    Chile/Continental: غرينتش-٣ (توقيت تشيلي الصيفي)
  </option>
  
      
  <option
    value="EST"
  >
    EST: غرينتش-٥ (التوقيت الرسمي الشرقي لأمريكا الشمالية)
  </option>
  
      
  <option
    value="Etc/GMT-7"
  >
    Etc/GMT-7: غرينتش+٧ (غرينتش+٠٧:٠٠)
  </option>
  
      
  <option
    value="Europe/Brussels"
  >
    Europe/Brussels: غرينتش+١ (توقيت وسط أوروبا الرسمي)
  </option>
  
      
  <option
    value="GMT"
  >
    GMT: UTC (التوقيت العالمي المنسق)
  </option>
  
      
  <option
    value="Indian/Maldives"
  >
    Indian/Maldives: غرينتش+٥ (توقيت جزر المالديف)
  </option>
  
      
  <option
    value="Pacific/Tarawa"
  >
    Pacific/Tarawa: غرينتش+١٢ (توقيت جزر جيلبرت)
  </option>
  
      
  <option
    value="UTC"
  >
    UTC: UTC (التوقيت العالمي المنسق)
  </option>
  
      
  <option
    value="Universal"
  >
    Universal: UTC (التوقيت العالمي المنسق)
  </option>
  
      
  <option
    value="Zulu"
  >
    Zulu: UTC (التوقيت العالمي المنسق)
  </option>
  
    
</select>
`;
